File: lib/compiler/macro_expander.dart
import 'package:bliss/compiler/node.dart';

class _MacroDefinition {
  final List<String> replacementArgs;
  final List<Node> bodyNodes;
  _MacroDefinition({required this.replacementArgs, required this.bodyNodes});

  @override
  String toString() {
    return "$runtimeType(replacementArgs: $replacementArgs, bodyNodes: $bodyNodes)";
  }
}

class MacroExpander {
  final _macroDefinitions = <String, _MacroDefinition>{};

  void _registerMacroDefinitions(List<Node> nodes) {
    final macroDefinitionNodesToRemove = <ListNode>[];

    for (final node in nodes) {
      switch (node) {
        case ListNode(
              contents: [
                SymbolNode(contents: "defmacro"),
                SymbolNode(contents: final String name),
                ValueListNode(contents: final List<Node> args),
                ...final List<Node> bodyNodes,
              ],
            )
            when args.isNotEmpty && args.every((x) => x is SymbolNode):
          _macroDefinitions[name] = _MacroDefinition(
            replacementArgs: args.map((x) => x.contents as String).toList(),
            bodyNodes: bodyNodes,
          );
          macroDefinitionNodesToRemove.add(node);

          _registerMacroDefinitions(node.contents);
        case ListNode(contents: [SymbolNode(contents: "defmacro"), ...]):
          throw "Invalid macro definition $node";
        case ListNode():
          _registerMacroDefinitions(node.contents);
        default:
          null;
      }
    }

    // Removes any macro definition nodes from the ast as they are not of use to us anymore and would be unnecessary to send to the emitter etc.
    for (var nodeToRemove in macroDefinitionNodesToRemove) {
      nodes.remove(nodeToRemove);
    }
  }

  void _expandMacroUsages(List<Node> nodes) {
    (List<Node>, bool) expandMacroUsage(
      String name,
      List<Node> args,
      List<String> replacementArgs,
      List<Node> bodyNodes,
    ) {
      var hasExpanded = false;

      final expandedBodyNodes =
          bodyNodes.map((macroBodyNode) {
            switch (macroBodyNode) {
              // An expansion is happening
              case SymbolNode(contents: final String name)
                  when replacementArgs.contains(name):
                hasExpanded = true;

                final replacementIdx = replacementArgs.indexOf(name);
                return args[replacementIdx];
              // We have a list so we need to go deeper
              case ListNode(contents: final List<Node> contents):
                final (
                  expandedContents,
                  didExpandInInnerList,
                ) = expandMacroUsage(name, args, replacementArgs, contents);

                if (!hasExpanded && didExpandInInnerList) hasExpanded = true;

                return ListNode(expandedContents);
              // We have a value list so we need to go deeper
              case ValueListNode(contents: final List<Node> contents):
                final (
                  expandedContents,
                  didExpandInInnerList,
                ) = expandMacroUsage(name, args, replacementArgs, contents);

                if (!hasExpanded && didExpandInInnerList) hasExpanded = true;

                return ValueListNode(expandedContents);
              // No match found, we dont do anything
              default:
                return macroBodyNode;
            }
          }).toList();

      return (expandedBodyNodes, hasExpanded);
    }

    for (final node in nodes) {
      var didExpandInLastAttempt = false;

      do {
        didExpandInLastAttempt = false;

        switch (node) {
          case ListNode(
                contents: [
                  SymbolNode(contents: final name),
                  ...final List<Node> args,
                ],
              )
              when _macroDefinitions.containsKey(name):
            final macroDefinition = _macroDefinitions[name]!;
            final (expandedBodyNodes, didExpandInAttempt) = expandMacroUsage(
              name,
              args,
              macroDefinition.replacementArgs,
              macroDefinition.bodyNodes,
            );

            didExpandInLastAttempt = didExpandInAttempt;

            if (macroDefinition.bodyNodes.length == 1 &&
                macroDefinition.bodyNodes[0] is ListNode) {
              node.contents = (expandedBodyNodes[0] as ListNode).contents;
            } else {
              node.contents = expandedBodyNodes;
            }
          case ListNode() || ValueListNode():
            _expandMacroUsages(node.contents);
          default:
        }
      } while (didExpandInLastAttempt);
    }
  }

  List<Node> expand(List<Node> astNodes) {
    _registerMacroDefinitions(astNodes);
    print("MACROS: $_macroDefinitions");

    _expandMacroUsages(astNodes);

    print("EXPANDED AST:");
    for (final node in astNodes) {
      print(node.format());
    }

    return astNodes;
  }
}

File: src/Compiler/MacroExpander.hs
{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}

module Compiler.MacroExpander (expand) where

import Compiler.Mod (Node (ListNode, SymbolNode, ValueListNode), isSymbolNode)
import Data.List (elemIndex, uncons)
import Data.Maybe (fromMaybe, isJust, maybeToList)
import Utils (fixedPoint)

data MacroDefinition = MacroDefinition
  { replacementArgs :: [String],
    macroNodes :: [Node]
  }
  deriving (Show)

registerMacroDefinitions :: [Node] -> ([(String, MacroDefinition)], [Node])
registerMacroDefinitions [] = ([], [])
registerMacroDefinitions (node : rest) =
  let -- (defmacro name [...] ...)
      processNode (ListNode (SymbolNode "defmacro" : SymbolNode name : ValueListNode args : bodyNodes))
        | not $ null args && all isSymbolNode args =
            let stringArgs = map (\(SymbolNode s) -> s) args
                newMacroDefinition = (name, MacroDefinition {replacementArgs = stringArgs, macroNodes = bodyNodes})
             in ([newMacroDefinition], Nothing)
      -- invalid macro
      processNode (ListNode (SymbolNode "defmacro" : _)) = error $ "Invalid macro definition: " ++ show node
      -- continue downwards to find more macros
      processNode (ListNode childNodes) =
        let (macrosFromChildren, remainingChildren) = registerMacroDefinitions childNodes
         in (macrosFromChildren, Just (ListNode remainingChildren))
      -- anything else
      processNode node' =
        ([], Just node')
      (macrosFromNode, maybeNode) = processNode node
      (macrosFromRest, nodesFromRest) = registerMacroDefinitions rest
   in (macrosFromNode ++ macrosFromRest, maybeToList maybeNode ++ nodesFromRest)

expandMacroUsagesPass :: [(String, MacroDefinition)] -> [Node] -> [Node]
expandMacroUsagesPass macroDefinitions astNodes =
  let -- found variadic arg, expand it
      expandArg macroDefinition args (SymbolNode ('&' : argName))
        | ("&" ++ argName) `elem` replacementArgs macroDefinition =
            let replacementIdx = fromMaybe (-1) $ elemIndex ("&" ++ argName) $ replacementArgs macroDefinition
                (_, replacementNodes) = splitAt replacementIdx args
             in ListNode $ SymbolNode "do" : replacementNodes -- wrap with (do ...) just in case
            -- found arg, expand it
      expandArg macroDefinition args (SymbolNode argName)
        | argName `elem` replacementArgs macroDefinition =
            let replacementIdx = fromMaybe (-1) $ elemIndex argName $ replacementArgs macroDefinition
                replacementNode = args !! replacementIdx
             in replacementNode
      -- continue deeper to find more args to expand
      expandArg macroDefinition args (ListNode exprs) = ListNode $ map (expandArg macroDefinition args) exprs
      expandArg macroDefinition args (ValueListNode exprs) = ValueListNode $ map (expandArg macroDefinition args) exprs
      -- anything else, dont do anything
      expandArg _ _ node = node
      isLastArgVariadic args = case uncons (last args) of
        Just ('&', _) -> True
        _ -> False
      -- found a macro usage, expand it node by node
      processNode (ListNode (SymbolNode name : args))
        | isJust $ lookup name macroDefinitions =
            let defaultMacroDefinition = (MacroDefinition {replacementArgs = [], macroNodes = []})
                macroDefinition = fromMaybe defaultMacroDefinition $ lookup name macroDefinitions
                newNode = ListNode $ SymbolNode "do" : macroNodes macroDefinition -- wrap with (do ...) just in case
                expandedNode = expandArg macroDefinition args newNode
                -- validate args to be the defined amount
                definedArgs = replacementArgs macroDefinition
                numDefinedArgs = length definedArgs
                numProvidedArgs = length args
                isVariadic = isLastArgVariadic definedArgs
             in case () of
                  _
                    | isVariadic && numProvidedArgs >= numDefinedArgs -> expandedNode
                    | not isVariadic && numProvidedArgs == numDefinedArgs -> expandedNode
                    | otherwise -> error $ "Invalid macro usage: " ++ name ++ " " ++ show args
      -- go deeper to find more macro definitions
      processNode (ListNode exprs) = ListNode $ map processNode exprs
      processNode (ValueListNode exprs) = ValueListNode $ map processNode exprs
      -- anything else, do nothing
      processNode node = node
   in map processNode astNodes

expandMacroUsages :: [(String, MacroDefinition)] -> [Node] -> [Node]
expandMacroUsages macroDefinitions astNodes =
  let expandPass = expandMacroUsagesPass macroDefinitions
   in fixedPoint expandPass astNodes

expand :: [Node] -> [Node]
expand astNodes =
  let (macroDefinitions, nonMacroDefinitionNodes) = registerMacroDefinitions astNodes
      expandedAst = expandMacroUsages macroDefinitions nonMacroDefinitionNodes
   in expandedAst